ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java

```diff
@@ -106,16 +108,31 @@ public void run(MLAgent mlAgent, Map<String, String> params, ActionListener<Obje
         List<MLToolSpec> toolSpecs = mlAgent.getTools();
         String memoryType = mlAgent.getMemory().getType();
         String memoryId = params.get(MEMORY_ID);
+        // for regenerate, original interaction id must provide
+        String regenerateInteractionId = params.get(REGENERATE_INTERACTION_ID);
         String appType = mlAgent.getAppType();
         String title = params.get(QUESTION);
+        if (null != regenerateInteractionId && memoryId == null) {
+            listener.onFailure(new MLValidationException("memory id must provide for regenerate"));
+            return;
+        }
 
         ConversationIndexMemory.Factory conversationIndexMemoryFactory = (ConversationIndexMemory.Factory) memoryFactoryMap.get(memoryType);
         conversationIndexMemoryFactory.create(title, memoryId, appType, ActionListener.<ConversationIndexMemory>wrap(memory -> {
             memory.getMessages(ActionListener.<List<Interaction>>wrap(r -> {
                 List<Message> messageList = new ArrayList<>();
                 Iterator<Interaction> iterator = r.iterator();
+                boolean interactionIdFound = false;
                 while (iterator.hasNext()) {
                     Interaction next = iterator.next();
+                    if (next.getId().equals(regenerateInteractionId)) {
+                        log.info("Regenerate for existing interaction {}", regenerateInteractionId);
+                        interactionIdFound = true;
+                        // if no new question provided, use original question
+                        params.computeIfAbsent(QUESTION, key -> next.getInput());
+                        // there may have other new interactions happened after this interaction
```

Comment:
Will question be specified in case of re-generation request? if so why would it be different from previous interaction question?

Reply:
For current UI design, question will not been passed as parameter when regenerate, it will use original one. But think what chatGPT and other chatbot does, user could edit the question and regenerate, to support that as well, adding this logic here.

if user don't have specified the question, use original one, otherwise use new one.